### PR TITLE
Add IPv6 support

### PR DESCRIPTION
**Pull request by p0lycarpio**
[linuxserverurl]: https://linuxserver.io
[linuxserver.io][linuxserverurl]

------------------------------

 - [x] I have read the [contributing](https://github.com/linuxserver/docker-duckdns/blob/master/.github/CONTRIBUTING.md) guideline and understand that I have made the correct modifications

------------------------------

## Description:

This PR adds the ability to update one or more DuckDNS domains with the device's public IPv6 address.

## Benefits of this PR and context:

The current script only updates the public IPv4 address through DuckDNS autodetection. Where DuckDNS [does not yet know how to determine the public IPv6 address](https://www.duckdns.org/faqs.jsp#:~:text=Q%3A%20why%20can%27t%20you%20detect%20IPv6%20addresses%3F) of a device, it is possible to provide it with the address as a [parameter](https://www.duckdns.org/spec.jsp#:~:text=ip%20%2D%20OPTIONAL%20%2D%20if%20left%20blank%20we%20detect%20IPv4%20addresses%2C%20if%20you%20want%20you%20can%20supply%20a%20valid%20IPv4%20or%20IPv6%20address%0Aipv6%20%2D%20OPTIONAL%20%2D%20a%20valid%20IPv6%20address%2C%20if%20you%20specify%20this%20then%20the%20autodetection%20for%20ip%20is%20not%20used).

The `UPDATE_BOTH` and `FORCE_IPV6` variables are used to retrieve the IPv6 address of the current device and to update the DNS record.

## How Has This Been Tested?

The Docker image was built locally and tested with either one of two variables set to true.

## Source / References:

Related to #39 and #53

### CODE REVIEW

**Comment by ReneHezser** [comment]
I'd love to have IPV6 support. Will this PR be merged?

**Comment by aptalca** [comment]
Adding 2 new env vars for this is unnecessary. It would need to be combined into one var, defaulting to ipv4 only.

**Comment by ReneHezser** [comment]
I agree. @p0lycarpio would you modify your code to use only one variable?
If you don't have time, I can do so.

**Comment by p0lycarpio** [comment]
Ok @aptalca @ReneHezser, I've removed the first variable. Everything looks good?

**Comment by aptalca** [comment]
Why not a single var that lets the user select between `ipv4` (default), `ipv6` and `both`?

**Comment by ReneHezser** [comment]
[like]  René Hézser reacted to your message:
________________________________

File: Dockerfile
Comment (aptalca) [inline Dockerfile]: I wouldn't recommend this. Portainer and synology gui read this and externally set it on an updated container: https://info.linuxserver.io/issues/2023-05-22-portainer/

It's better to modify the if block in the init script so the `ipv6` and `both` can be the first two tests (`if`/`elif`), and `else` would do ipv4 only. That way ipv4 is truly the default option, so even when the user sets it to something invalid, the code will treat it as `ipv4`.

Comment (p0lycarpio) [reply Dockerfile]: Right. I did not know that... I fix

**Comment by aptalca** [comment]
Looks good, thanks. Please add a changelog entry to readme-vars.yml and we'll merge

**Comment by p0lycarpio** [comment]
Hi @aptalca what's new on this PR? Are there things missing for merge?

**Comment by aptalca** [comment]
Internal discussion on which service(s) to use to detect external ip and whether to provide choice to end user due to privacy related concerns

**Comment by ReneHezser** [comment]
It would be a shame if this feature won't make it into the image.

**Comment by gucki** [comment]
> Internal discussion on which service(s) to use to detect external ip and whether to provide choice to end user due to privacy related concerns

It's really sad this isn't supported yet. Why not just go for a simple solution and people who are afraid of their privacy can just disable it until a "safe solution" or "choice" has been implemented...

**Comment by p0lycarpio** [comment]
That does not make sense. The IPv6 option is disabled by default so I don't see any privacy issues. At a pinch, we could add in the documentation an explanation that the activation of IPv6 uses an external service. Alternatively, we could vary the URL of the service to use, so that the user can specify the IPv6 discovery service to use...

**Comment by CAMOBAP** [comment]
@aptalca thanks for your hard work

Looks like all feedback addressed sofar, it will be nice to see the functionality in the image soon

**Comment by nemchik** [comment]
Just to clarify, the main concern here is that the current behavior only communicates between your container and the duckdns service. This method of IP detection includes an additional party. Even if it is not used by default, a user might decide they want to use the `ipv6` or `both` option and be unaware that their IP is being detected by using a service that is not duckdns. Unfortunately I am unaware of a way to determine one's own WAN IP address without querying a 3rd party for a response. I would be more satisfied with something that uses a local resolver for this, but I don't know if that's possible.

In the very least, using a known reputable provider would be the best option. I know Cloudflare and Google are both options, but there are some who would prefer not to trust them either.

CF would use
```
dig +short ch txt whoami.cloudflare -4 @1.1.1.1
dig +short ch txt whoami.cloudflare -6 @2606:4700:4700::1111
```

**Comment by drizuid** [comment]
just pop it in the readme; as the only ipv6 member of our ipv4 cult, i think this is a good pr :)

**Comment by nemchik** [comment]
> just pop it in the readme; as the only ipv6 member of our ipv4 cult, i think this is a good pr :)

i also have ipv6 enabled (although being honest, i barely understand it and use it more for testing stuff works with it).

anyway, i like the PR, i like the syntax. I'd like to see a readme section specifically explaining that an external service is used for detection if the variable has any value. also, it would make sense that if the value is explicitly set to ipv4, use the external detection, but if the value is not set at all then use the original method where duckdns detects the address.

**Comment by drizuid** [comment]
> > just pop it in the readme; as the only ipv6 member of our ipv4 cult, i think this is a good pr :)

> anyway, i like the PR, i like the syntax. I'd like to see a readme section specifically explaining that an external service is used for detection if the variable has any value. also, it would make sense that if the value is explicitly set to ipv4, use the external detection, but if the value is not set at all then use the original method where duckdns detects the address.

@p0lycarpio can you take care of the items nemchik listed above? Obviously we delayed you until october so you get some credit for hacktoberfest, this was all by design.

**Comment by p0lycarpio** [comment]
Hi @nemchik @drizuid,
Here are the changes made based on your feedback.
I'll let you take a look and I'd be happy if you can add the label "hacktoberfest-accepted" :)

**Comment by drizuid** [comment]
> I'd be happy if you can add the label "hacktoberfest-accepted" :)

no need, the whole repo is tagged hacktoberfest

I'll let nemchik do the review though, just be patient :)

File: readme-vars.yml
Comment (nemchik) [inline readme-vars.yml]:
```suggestion
  - { env_var: "UPDATE_IP", env_value: "ipv4", desc: "Set to `ipv6` or `ipv4` to update  **only** your public IPv4/6 address. Set to `both` to update IPv6 and IPv4 address. This variable makes use of a [third-party service](#notice-regarding-automatic-detection). Omitting this variable uses DuckDNS for detection and only supports IPv4."}
```

File: readme-vars.yml
Comment (nemchik) [inline readme-vars.yml]:
```suggestion
  ## Notice regarding automatic detection

  Using the `UPDATE_IP` variable whatever its value (`ipv4`, `ipv6` or `both`) uses external *Cloudflare whoami* service to detect public IP addresses.
  **Be aware that using this variable will query a third-party service other than DuckDNS.**

  Omitting the `UPDATE_IP` variable uses DuckDNS for detection and only supports IPv4.

```

**Review by nemchik — CHANGES_REQUESTED** [review]
This looks great. I added some suggested edits just for wording and clarity and to slightly shorten the anchor link. I'll approve once the review is addressed.

Awesome work on this!

**Comment by nemchik** [comment]
> I am a bot, here are the test results for this PR: https://ci-tests.linuxserver.io/lspipepr/duckdns/256b24bf-pkg-256b24bf-dev-10a1bc05554dcd30dd35666399ce69ede08e35b6-pr-61/index.html https://ci-tests.linuxserver.io/lspipepr/duckdns/256b24bf-pkg-256b24bf-dev-10a1bc05554dcd30dd35666399ce69ede08e35b6-pr-61/shellcheck-result.xml
> 
> Tag	Passed
> amd64-256b24bf-pkg-256b24bf-dev-10a1bc05554dcd30dd35666399ce69ede08e35b6-pr-61	✅
> arm64v8-256b24bf-pkg-256b24bf-dev-10a1bc05554dcd30dd35666399ce69ede08e35b6-pr-61	✅

@p0lycarpio I've just tried testing the PR build from our bot. 

- Including the var, but leaving the value blank or setting it to a random string other than `ipv4`, `ipv6`, or `both` works as expected (uses duckdns detection)
- Not including the var works as expected (uses duckdns detection)
- Including the var set to `ipv4` works as expected (uses cf detection)
- Including the var set to `ipv6` or `both` **does not work**.

```
Something went wrong, please check your settings Sun Oct 15 13:15:59 CDT 2023
The response returned was:
```

And a blank line after that (so, no response returned?)

My first guess is the ipv6 needs to be url encoded, but I'm not positive.

Also now that I've used it, I would like to see the detection method echo'd to the log (I'll add a review with request for changes with what I mean shortly).

File: root/app/duck.sh
Comment (nemchik) [inline root/app/duck.sh]:
```suggestion
    # Use cloudflare to autodetect IP if UPDATE_IP is set
    if [ "${UPDATE_IP}" = "ipv4" ]; then
        echo "Detecting IPv4 via CloudFlare"
        IPV4=$(dig +short ch txt whoami.cloudflare -4 @1.1.1.1 | sed 's/"//g')
        RESPONSE=$(curl -sS --max-time 60 "https://www.duckdns.org/update?domains=${SUBDOMAINS}&token=${TOKEN}&ip=${IPV4}")
    elif [ "${UPDATE_IP}" = "ipv6" ]; then
        echo "Detecting IPv6 via CloudFlare"
        IPV6=$(dig +short ch txt whoami.cloudflare -6 @2606:4700:4700::1111 | sed 's/"//g')
        RESPONSE=$(curl -sS --max-time 60 "https://www.duckdns.org/update?domains=${SUBDOMAINS}&token=${TOKEN}&ip=${IPV6}")
    elif [ "${UPDATE_IP}" = "both" ]; then
        echo "Detecting IPv4 and IPv6 via CloudFlare"
        IPV4=$(dig +short ch txt whoami.cloudflare -4 @1.1.1.1 | sed 's/"//g')
        IPV6=$(dig +short ch txt whoami.cloudflare -6 @2606:4700:4700::1111 | sed 's/"//g')
        RESPONSE=$(curl -sS --max-time 60 "https://www.duckdns.org/update?domains=${SUBDOMAINS}&token=${TOKEN}&ip=${IPV4}&ipv6=${IPV6}")
    else
    # Use DuckDns to autodetect IPv4 (default behaviour)
        echo "Detecting IPv4 via DuckDNS"
        RESPONSE=$(curl -sS --max-time 60 "https://www.duckdns.org/update?domains=${SUBDOMAINS}&token=${TOKEN}&ip=")
    fi

```

**Review by nemchik — CHANGES_REQUESTED** [review]

**Comment by p0lycarpio** [comment]
Hi @nemchik,
Even though the Duckdns API specification states that the "ip" parameter can contain an IPv4 or IPv6 address, I have managed to reproduce tests where I get the return "KO" when passing IPv6 as "ip". I made the changes in IPv6 condition.

As for the dig command, I think it returns nothing when the machine does not have an IPv6 address or cannot reach the outside via IPv6. The query executes 3 times before timeout.

```sh
Detecting IPv4 and IPv6 via CloudFlare
;; UDP setup with 2606:4700:4700::1111#53(2606:4700:4700::1111) for whoami.cloudflare failed: network unreachable.
;; UDP setup with 2606:4700:4700::1111#53(2606:4700:4700::1111) for whoami.cloudflare failed: network unreachable.
;; UDP setup with 2606:4700:4700::1111#53(2606:4700:4700::1111) for whoami.cloudflare failed: network unreachable.
Something went wrong, please check your settings Thu Oct 19 16:32:15 CEST 2023
The response returned was:

```

I tried several methods, like with the "host" network or adding the NET_ADMIN and NET_RAW rules with cap-add or changing the [Docker daemon configuration](https://docs.docker.com/config/daemon/ipv6/) but none worked...

**Comment by nemchik** [comment]
Using host network mode should work for ipv6 detection. I'm sure there are other ways to do it, but I believe it requires making docker setup containers with ipv6. This may be a topic for @drizuid to chime in.

**Comment by drizuid** [comment]
I've never used host for any of this, i create a custom docker bridge and use a carved out /112 for docker. I give the containers public IPs from that /112 and just route the /112 from my edge to the docker host. HOWEVER, i do not use duckdns nor any other automated DDNS container, i just have a small bash script on cron that updates my IPs. In my case, this is needed because where the duckdns container is usually updating one A record that has a bunch of cnames pointing to it with your single public ip, my containers each have their own public IPv6 address, so i must update multiple A records individually (the script has only run once in 3 yrs because that is the only time my IPv6 block ever changed (transition from gpon to xgs-pon) 

That said, my answer probably doesnt help with the immediate issue, but it might shine a light on "proper" (IMO) use of ipv6

**Comment by p0lycarpio** [comment]
That's right, it also works on my side in host mode.
As for which method to use, creating a new bridge network is a good idea and should be used on all containers. Even if you don't need to update the IPv6 address in the container often, I don't see how this can be automated and what it implies for end users.

In my opinion, the host network is simpler and less likely to break. The fact that this container doesn't expose any ports and the services contacted are known, can this justify non-isolated network usage?

**Comment by nemchik** [comment]
> In my opinion, the host network is simpler and less likely to break. The fact that this container doesn't expose any ports and the services contacted are known, can this justify non-isolated network usage?

Yes, but we should include a note in the readme about it.

**Review by nemchik — APPROVED** [review]
Tested and working on my end. Thanks for all your work on this!

**Pull request merged**
